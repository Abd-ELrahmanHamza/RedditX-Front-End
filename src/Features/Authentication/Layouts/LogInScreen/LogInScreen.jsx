/* eslint-disable no-unused-vars */
/* eslint-disable react-hooks/exhaustive-deps */
import { useEffect } from "react";

import FacebookLogin from "react-facebook-login";

import FormInput from "Features/Authentication/Components/FormInput/FormInput";

import Button from "Features/Authentication/Components/Button/Button";

import { FcGoogle } from "react-icons/fc";
import { BsFacebook } from "react-icons/bs";

import {
  signInWithGooglePopup,
  signInWithFacebookPopup,
} from "Features/Authentication/Utils/Firebase";

import {
  AuthContainer,
  ButtonsContainer,
  Privacy,
  ErrorParagraph,
  SignInWithGoogle,
  SignInWithFacebook,
  OrHeader,
  AuthParagraph,
  AuthHeader,
  Forget,
  AuthContainerDiv,
} from "./LogInScreen.styled";

const USER_REGEX = /^[A-z][A-z0-9-_]{3,20}$/;
const PWD_REGEX = /^[A-z][A-z0-9-_]{8,20}$/;

/**
 * LoginScreen component that is used in Login Component
 * @param {Object} formFields Object contain the values of input fields
 * @param {Function} setFormFields Function to update the values of input fields
 * @param {boolean} validName Prop to know if the userName is valid or not
 * @param {Function} setValidName Function to set the validity of the userName
 * @param {boolean} validPassword Prop to know if the password is valid or not
 * @param {Function} setValidPassword Function to set the validity of the password
 * @param {boolean} userNameScreen Prop to know if we are in the userName screen or not
 * @param {Function} setUserNameScreen Function to set the state of userNameScreen
 * @param {boolean} passwordScreen Prop to know if we are in the password screen or not
 * @param {Function} setPasswordScreen Function to set the state of passwordScreen
 * @param {boolean} initialFocus Prop to know if the user made at least one focus on the input field or not
 * @param {Function} setInitialFocus Function to set the state of initialFocus
 * @param {Function} setErrMsg Function to set the error message
 * @returns {React.Component}  LoginScreen component that is used in Login Component
 */

const LogInScreen = ({
  formFields,
  setFormFields,
  validName,
  setValidName,
  validPassword,
  setValidPassword,
  userNameScreen,
  setUserNameScreen,
  passwordScreen,
  setPasswordScreen,
  initialFocus,
  initialFocus2,
  setInitialFocus,
  setInitialFocus2,
  setErrMsg,
}) => {
  const { userName, password } = formFields;

  /**
   * useEffect for userName field to check if the userName that the user entered is valid or not
   */
  useEffect(() => {
    if (userName.length > 0) {
      setInitialFocus(false);
    }
    setValidName(USER_REGEX.test(userName));
  }, [userName]);

  /**
   * useEffect for password field to check if the userName that the user entered is valid or not
   */
  useEffect(() => {
    if (password.length > 0) {
      setInitialFocus(false);
    }
    setValidPassword(PWD_REGEX.test(password));
  }, [password]);

  /**
   * Function to handle the submit of the form of login
   * @param {*} event
   */
  const handleSubmit = async (event) => {
    event.preventDefault();
  };

  /**
   * Function to handle any change on the input field of the login form (check if the userName or the email or the password is valid or not)
   * @param {*} event
   */

  const handleChange = (event) => {
    const { name, value } = event.target;
    if (name === "userName") {
      setValidName(USER_REGEX.test(userName));
    }
    if (name === "password") {
      setValidPassword(PWD_REGEX.test(password));
    }
    setFormFields({ ...formFields, [name]: value });
  };

  /**
   * Function to handle the response coming from sign in with google
   * @param {*} response
   */
  const handleCallbackResponse = (response) => {};

  /**
   * Adding some configurations to the signIn with google feature
   */

  useEffect(() => {
    /* global google */

    google.accounts.id.initialize({
      client_id:
        "598360538255-siv9rljce260ek4mvdsu2fk63h3u1g3b.apps.googleusercontent.com",
      callback: handleCallbackResponse,
    });

    google.accounts.id.renderButton(document.getElementById("signInDiv"), {
      theme: "outline",
      size: "large",
    });
  }, []);

  /**
   * Function to handle the response coming from sign in with facebook
   * @param {*} response
   */
  const responseFacebook = (response) => {};

  const logGoogleUser = async () => {
    const { user } = await signInWithGooglePopup();
  };
  const logFacebookUser = async () => {
    const { user } = await signInWithFacebookPopup();
   // console.log(user);
  };

  return (
    <>
      {
        <AuthContainer userNameScreen={userNameScreen}>
          <AuthContainerDiv userNameScreen={userNameScreen || passwordScreen}>
            <AuthHeader>Log In</AuthHeader>
            <AuthParagraph>
              By continuing, you agree are setting up a Reddit account and agree
              to our <Privacy>User Agreement</Privacy> and{" "}
              <Privacy>Privacy Policy</Privacy>.
            </AuthParagraph>
            <SignInWithGoogle>
              {/* <div id="signInDiv"></div> */}
              <button onClick={() => logGoogleUser()}>
                <FcGoogle size={22} />
                <span> Continue With Google</span>
              </button>
            </SignInWithGoogle>

            <SignInWithFacebook>
              <button onClick={() => logFacebookUser()}>
                <BsFacebook size={22} />
                <span> Continue With Facebook</span>
              </button>
            </SignInWithFacebook>

            <OrHeader>
              <span>OR</span>
            </OrHeader>

            <form onSubmit={handleSubmit}>
              <FormInput
                valid={validName}
                initialFocus={initialFocus}
                showIcon={true}
                label="Username"
                type="text"
                onChange={handleChange}
                name="userName"
                value={userName}
                onFocus={() => {
                  setInitialFocus(false);
                }}
              />

              {/* Show error message if the userName is not valid and the user made a focus on the it's input field */}
              <ErrorParagraph valid={validName || initialFocus}>
                Username must be between 3 and 20 characters
              </ErrorParagraph>

              <FormInput
                label="Password"
                valid={validPassword}
                initialFocus={initialFocus2}
                showIcon={true}
                type="password"
                onChange={handleChange}
                name="password"
                value={password}
                onFocus={() => {
                  setInitialFocus2(false);
                }}
              />

              <Forget>
                Forget your{" "}
                <button
                  onClick={() => {
                    setInitialFocus(true);
                    setUserNameScreen(true);
                    setPasswordScreen(false);

                    setErrMsg("Not a valid email address");
                  }}
                >
                  username
                </button>{" "}
                or{" "}
                <button
                  onClick={() => {
                    setInitialFocus(true);
                    setUserNameScreen(false);
                    setPasswordScreen(true);

                    setErrMsg("Not a valid email address");
                  }}
                >
                  password
                </button>
              </Forget>

              <ButtonsContainer>
                {/** Show an enabled button if the password and userName are valid */}
                {validName && validPassword && (
                  <Button
                    valid={validName && validPassword}
                    type="button"
                    onClick={() => {}}
                  >
                    Log In
                  </Button>
                )}

                {/** Show an disabled button if the password or userName are not valid */}

                {(!validName || !validPassword) && (
                  <Button
                    disabled
                    valid={validName && validPassword}
                    type="button"
                  >
                    Log In
                  </Button>
                )}
              </ButtonsContainer>
            </form>
          </AuthContainerDiv>
        </AuthContainer>
      }
    </>
  );
};

export default LogInScreen;
