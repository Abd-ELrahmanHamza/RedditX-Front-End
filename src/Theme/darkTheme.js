/**
 * dark theme colors
 *
 * @type {{id:string}}
 */
const darkTheme = {
  // The id of the theme
  id: "dark",
  // The canvas color is the background color of the website
  canvas: "#030303",
  // The colors of the text on the website
  color: {
    // The primary color of the text on the website
    primary: "#d7dadc",
    // The secondary color of the website looks like primary in dark theme
    secondary: "#d7dadc",
    // The color of the muted text on the website
    muted: "#7c7c7c",
    // The color of the links on the website
    link: "#0079d3",
    // The orange color of the website
    orange: "#d93a00",
    // The pink color of the website
    pink: "#fb133a",
    // The color of the text of the main links of search page
    titleText_of_link: "#D7DADC",

    // Some colors for authentication part
    logInButtonColorHovered: "#dc4a14",
    logInBlueHovered: "#3394dc",
    baseButtonColor: "#fff",
    subColor: "#797b7d",
    mainInputColor: "black",
    borderColor: "#86888a",
    blueValid: "#24a0ed",
    mainInputBackGroundColor: "#f6f7f8",
    initialBorderColor: "#e6e6e5",
    initialLabelColor: "#aeadad",
    errorColor: "#ea0027",
    orHeaderColor: "#93908e",
    authParagraphColor: "#1a1a1b",
    signInWithFacebookColor: "#237df3",
  },
  // The colors of the backgrounds in the website
  background: {
    // The primary color of the background on the website
    primary: "#1a1a1b",
    // The color of the muted background on the website
    muted: "#7c7c7c3d",
    // The pink color of the website
    pink: "#fb133a",
    // The light muted color of the website
    lightMuted: "#f6f7f8",
  },
  // Border colors
  borderColor: {
    // Primary border color
    primary: "#818384",
    // Secondary border color
    secondary: "#343536",
    // Muted border color
    muted: "#878a8c80",
    // The button hover color of the search page
    fill: "#2D2D2F",
    // The button active color of the search page
    active: "#272729",
    // button background color hover of the search page cards
    card_hover: "#353537",
    // Comment background color in search page
    comment: "#17232D",
  },
  // The colors of the lines in the website
  lineColor: {
    primary: "#343536",
    // Post Border Line Color Search Page
    PostBorder: "#343536",
  },
  // The colors of the buttons in the website
  button: {
    // The color when hover on buttons on the website
    muted: "#1a1a1b",
  },
};

export default darkTheme;
