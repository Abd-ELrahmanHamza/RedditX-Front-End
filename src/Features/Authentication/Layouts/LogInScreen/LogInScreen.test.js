import { shallow } from "enzyme";

import { screen, render, fireEvent, waitFor } from "@testing-library/react";
import { ThemeProvider } from "styled-components";

import darkTheme from "Theme/darkTheme";

import { BrowserRouter } from "react-router-dom";

import LogInScreen from "./LogInScreen";




const defaultFormFields = {
  userName: "",
  password: "",
  email: "",
};

const validPassword = false;
const validName = false;
const userNameScreen = false;
const passwordScreen = false;
const initialFocus = false;
const initialFocus2 = false;
const setValidPassword = jest.fn();
const setModalShowLogIn = jest.fn();
const setFormFields = jest.fn();
const setUserNameScreen = jest.fn();
const setValidName = jest.fn();
const setPasswordScreen = jest.fn();
const setInitialFocus = jest.fn();
const setInitialFocus2 = jest.fn();
const setErrMsg = jest.fn();

const mockedUsedNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUsedNavigate,
}));

describe("LogInScreen Layout", () => {
  const wrapper = shallow(
    <BrowserRouter>
      <LogInScreen
        setFormFields={setFormFields}
        formFields={defaultFormFields}
        validName={validName}
        setValidName={setValidName}
        validPassword={validPassword}
        setValidPassword={setValidPassword}
        userNameScreen={userNameScreen}
        setUserNameScreen={setUserNameScreen}
        passwordScreen={passwordScreen}
        setPasswordScreen={setPasswordScreen}
        initialFocus={initialFocus}
        initialFocus2={initialFocus2}
        setInitialFocus={setInitialFocus}
        setInitialFocus2={setInitialFocus2}
        setModalShowLogIn={setModalShowLogIn}
        setErrMsg={setErrMsg}
      />
    </BrowserRouter>
  );

  const userNameInput = wrapper.childAt(0).dive().find('[id="login-username"]');
  const userNameError = wrapper.childAt(0).dive().find('[id="username-error"]');
  const passwordInput = wrapper.childAt(0).dive().find('[id="login-password"]');
  const forgetUserName = wrapper
    .childAt(0)
    .dive()
    .find('[id="forgetUserName"]');
  const forgetPassword = wrapper
    .childAt(0)
    .dive()
    .find('[id="forgetPassword"]');


  it("this is a test for userName field", () => {
    userNameInput.simulate("change", {
      target: { name: "userName", value: "g" },
    });
    expect(userNameError.text()).toEqual(
      "Username must be between 3 and 20 characters"
    );

  

    expect(setValidName).toBeCalled();
    expect(setInitialFocus).toBeCalled();
    expect(setInitialFocus2).not.toBeCalled();
  });

  it("this is a test for password field", () => {
    passwordInput.simulate("change", {
      target: { name: "password", value: "g" },
    });
    expect(setValidPassword).toBeCalled();
    expect(setInitialFocus2).toBeCalled();
    expect(setInitialFocus).not.toBeCalled();
  });

  it("this is a test for forgetUserName link", () => {
    forgetUserName.simulate("click");
    expect(setInitialFocus).toBeCalled();
    expect(setPasswordScreen).toBeCalled();
    expect(setErrMsg).toBeCalled();
    expect(setUserNameScreen).toBeCalled();
  });

  it("this is a test for forgetPassword link", () => {
    forgetPassword.simulate("click");
    expect(setInitialFocus).toBeCalled();
    expect(setPasswordScreen).toBeCalled();
    expect(setErrMsg).toBeCalled();
    expect(setUserNameScreen).toBeCalled();
  });



  test("username-error should be found", async () => {
    render(
      <BrowserRouter>
        <ThemeProvider theme={darkTheme}>
        <LogInScreen
        setFormFields={setFormFields}
        formFields={defaultFormFields}
        validName={validName}
        setValidName={setValidName}
        validPassword={validPassword}
        setValidPassword={setValidPassword}
        userNameScreen={userNameScreen}
        setUserNameScreen={setUserNameScreen}
        passwordScreen={passwordScreen}
        setPasswordScreen={setPasswordScreen}
        initialFocus={initialFocus}
        initialFocus2={initialFocus2}
        setInitialFocus={setInitialFocus}
        setInitialFocus2={setInitialFocus2}
        setModalShowLogIn={setModalShowLogIn}
        setErrMsg={setErrMsg}
      />
        </ThemeProvider>
      </BrowserRouter>
    );
    const userNameInput = screen.getByTestId("username");

    fireEvent.change(userNameInput, {
      target: { value: "M" },
    });

    userNameInput.focus();

    const userNameError = screen.getByTestId("username-error");

    await waitFor(() => {
      expect(userNameError).toBeVisible();
    });
  });
  

  test("password-input should be found", () => {
    render(
      <BrowserRouter>
        <ThemeProvider theme={darkTheme}>
        <LogInScreen
        setFormFields={setFormFields}
        formFields={defaultFormFields}
        validName={validName}
        setValidName={setValidName}
        validPassword={validPassword}
        setValidPassword={setValidPassword}
        userNameScreen={userNameScreen}
        setUserNameScreen={setUserNameScreen}
        passwordScreen={passwordScreen}
        setPasswordScreen={setPasswordScreen}
        initialFocus={initialFocus}
        initialFocus2={initialFocus2}
        setInitialFocus={setInitialFocus}
        setInitialFocus2={setInitialFocus2}
        setModalShowLogIn={setModalShowLogIn}
        setErrMsg={setErrMsg}
      />
        </ThemeProvider>
      </BrowserRouter>
    );
    const userNameInput = screen.getByTestId("password");
    expect(userNameInput).toBeInTheDocument();
  });
});
