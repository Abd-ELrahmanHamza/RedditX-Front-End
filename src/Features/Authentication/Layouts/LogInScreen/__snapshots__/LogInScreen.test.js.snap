// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogInScreen Layout this is a test for LogInScreen Layout 1`] = `ShallowWrapper {}`;
